### PR TITLE
expression: extract process of getting full cmp type to method.

### PR DESCRIPTION
**Pull request by winoros**
Since `in` function also need this to make result better, extract the this part to single method.
PTAL @XuHuaiyu @zz-jason

### CODE REVIEW

**Comment by hanfei1991** [comment]
Good Job !
LGTM

File: expression/builtin_compare.go [hunk 1]
```diff
@@ -847,6 +847,62 @@ func getCmpType(lhs, rhs types.EvalType, lft, rft *types.FieldType) types.EvalTy
 	return types.ETReal
 }
 
+func getCmpType(lhs, rhs Expression, moreAccure bool) types.EvalType {
```
Comment (zz-jason) [inline expression/builtin_compare.go]: how about set `getCmpType` to be a member function of `compareFunctionClass` ?

Comment (winoros) [reply expression/builtin_compare.go]: `in` will use this to check whether it has no need to be rewritten to `or`.

Comment (XuHuaiyu) [inline expression/builtin_compare.go]: 1. add a comment for this func
2. moreAccure is useless?

File: expression/builtin_compare.go [hunk 2]
```diff
@@ -946,75 +1002,13 @@ func (c *compareFunctionClass) getFunction(ctx context.Context, rawArgs []Expres
...
-			}
-			if col.GetType().Tp == mysql.TypeDuration {
-				sig, err = c.generateCmpSigs(ctx, args, types.ETDuration)
-			} else {
-				sig, err = c.generateCmpSigs(ctx, args, types.ETDatetime)
-			}
-		}
-	}
+	cmpType := getCmpType(args[0], args[1], true)
 	if err != nil {
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: can be removed.

File: expression/builtin_compare.go [hunk 3]
```diff
@@ -946,75 +1002,13 @@ func (c *compareFunctionClass) getFunction(ctx context.Context, rawArgs []Expres
...
-			sig, err = c.generateCmpSigs(ctx, args, types.ETReal)
-		}
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
+	sig, err = c.generateCmpSigs(ctx, args, cmpType)
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
 	return sig, nil
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: return sig, errors.Trace(err)

File: expression/builtin_compare.go [hunk 4]
```diff
@@ -946,75 +1002,13 @@ func (c *compareFunctionClass) getFunction(ctx context.Context, rawArgs []Expres
...
-		case types.ETDecimal:
-			sig, err = c.generateCmpSigs(ctx, args, types.ETDecimal)
-		case types.ETReal:
-			sig, err = c.generateCmpSigs(ctx, args, types.ETReal)
-		}
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
+	sig, err = c.generateCmpSigs(ctx, args, cmpType)
+	if err != nil {
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: ditto

**Pull request merged**